google_guest_agent/events/events.go

```diff
@@ -48,14 +49,65 @@ type Watcher interface {
 // Manager defines the interface between events management layer and the
 // core guest agent implementation.
 type Manager struct {
-	watchers    map[string]Watcher
+	// watcherEvents maps the registered watchers and their events.
+	watcherEvents []*WatcherEventType
+
+	// watchersMap is a convenient manager's mapping of registered watcher instances.
+	watchersMap map[string]bool
+
+	// watchersMutex protects the watchers map.
+	watchersMutex sync.Mutex
+
+	// removingWatcherEvents is a map of watchers being removed.
+	removingWatcherEvents map[string]bool
+
+	// running is a flag indicating if the Run() was previously called.
+	running bool
+
+	// runningMutex protects the running flag.
+	runningMutex sync.Mutex
+
+	// subscribers maps the subscribed callbacks.
 	subscribers map[string][]*eventSubscriber
+
+	// subscribersMutex protects subscribers member/map of the manager object.
+	subscribersMutex sync.Mutex
+
+	// queue queue struct manages the running watchers, when it gets to len()
+	// down to zero means all watchers are done and we can signal the other
+	// control go routines to leave(given we don't have any more job left to
+	// process).
+	queue *watcherQueue
 }
 
-// Config offers a mechanism for the consumers to configure the Manager behavior.
-type Config struct {
-	// Watchers lists the enabled watchers, of not provided all available watchers will be enabled.
-	Watchers []string
+// watcherQueue wraps the watchers <-> callbacks communication as well as the
+// communication/coordination of the multiple control go routine i.e. the one
+// responsible to calling callbacks after a event is produced by the watcher etc.
+type watcherQueue struct {
+	// queueMutex protects the access to watchersMap
+	queueMutex sync.Mutex
+
+	// watchersMap maps the currently running watchers.
+	watchersMap map[string]bool
+
+	// finishContextHandler is a channel used to communicate with the context handling
+	// go routine that it should finish/end its job (usually after all watchers are done).
+	finishContextHandler chan bool
+
+	// finishCallbackHandler is a channel used to communicate with the callback handling
+	// go routine that it should finish/end its job (usually after all watchers are done).
+	finishCallbackHandler chan bool
+
+	// watcherDone is a channel used to communicate that a given watcher is finished/done.
+	watcherDone chan string
+
+	// dataBus is the channel used to communicate between watchers (event producer) and the
+	// callback handler (event consumer managing go routine).
+	dataBus chan eventBusData
```

Comment:
We might not be super sensitive about latency and all, but being a channel of length 1 wouldn't all other watchers wait to write new response until previous one is processed? would this be a bottleneck?

Reply:
For simplicity this initial implementation is intended to have the callback handling/calling queued up so we don't have a go routine for each but a single go routine dequeueing the event processing. We can sure optimize it later if we have a need/requirement or if it becomes a bottleneck, in that case it would make sense to have a go routine per callback in the queue etc. Do you think we need this optimization now?